Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp

```diff
@@ -0,0 +1,323 @@
+                                                   const size_t nTimeIndexes,
+                                                   const size_t nBins)
+    : m_nDetectors(nDetectors), m_nTimeIndexes(nTimeIndexes), m_nBins(nBins),
+      m_histogram(BinEdges(nBins + 1, LinearGenerator(0.0, 1.0)),
+                  Counts(std::vector<double>(nBins, 0.0))),
+      m_indexingType(IndexingType::DEFAULT) {}
+
+void ScanningWorkspaceBuilder::setInstrument(
+    const boost::shared_ptr<const Geometry::Instrument> &instrument) {
+  if (instrument->getNumberDetectors() < m_nDetectors) {
+    throw std::logic_error("There are not enough detectors in the instrument "
+                           "for the number of detectors set in the scanning "
+                           "workspace builder.");
+  }
+
+  m_instrument = instrument;
+}
+
+/**
+ * Set a histogram to be used for all the workspace spectra. This can be used to
+ *set the correct bin edges, but only if the binning is identical for every
+ *spectra.
+ *
+ * @param histogram A histogram with bin edges defined
+ */
+void ScanningWorkspaceBuilder::setHistogram(
+    const HistogramData::Histogram &histogram) {
+  if ((histogram.counts().size() != m_nBins) ||
+      (histogram.binEdges().size() != m_nBins + 1))
+    throw std::logic_error(
+        "Histogram supplied does not have the correct size.");
+
+  m_histogram = histogram;
+}
+
+/**
+ * Set time ranges from a vector of start time, end time pairs.
+ *
+ * @param timeRanges A vector of DateAndTime pairs, corresponding to the start
+ *and end times
+ */
+void ScanningWorkspaceBuilder::setTimeRanges(const std::vector<
+    std::pair<Kernel::DateAndTime, Kernel::DateAndTime>> &timeRanges) {
+  verifyTimeIndexSize(timeRanges.size(), "start time, end time pairs");
+  m_timeRanges = timeRanges;
+}
+
+/**
+ * Set time ranges from a start time and a vector of durations
+ *
+ * @param startTime A DateAndTime object corresponding to the start of the first
+ *scan
+ * @param durations A vector of doubles containing the duration in seconds
+ */
+void ScanningWorkspaceBuilder::setTimeRanges(
+    const Kernel::DateAndTime &startTime,
+    const std::vector<double> &durations) {
+  verifyTimeIndexSize(durations.size(), "time durations");
+
+  std::vector<std::pair<Kernel::DateAndTime, Kernel::DateAndTime>> timeRanges =
+      {std::pair<Kernel::DateAndTime, Kernel::DateAndTime>(
+          startTime, startTime + durations[0])};
+
+  for (size_t i = 1; i < m_nTimeIndexes; ++i) {
+    const auto newStartTime = timeRanges[i - 1].second;
+    const auto endTime = newStartTime + durations[i];
+    timeRanges.push_back(std::pair<Kernel::DateAndTime, Kernel::DateAndTime>(
+        newStartTime, endTime));
+  }
+
+  setTimeRanges(timeRanges);
+}
+
+/**
+ * Supply a vector of vectors which contain positions. The inner vectors should
+ *contain the position for each time index, the outer vector the vector for each
+ *detector.
```

Comment:
So should we not also compare things like the outer vector position size with outer rotation vector size? There may be some other examples for this. I didn't see that as part of the validation.